### PR TITLE
feat: create Split Dialer

### PR DESCRIPTION
**Pull request by fortuna**

### CODE REVIEW

**Review by fortuna — COMMENTED** [review]
Thoughts?

File: transport/split/stream_dialer.go [hunk 1]
```diff
@@ -0,0 +1,46 @@
...
+)
+
+type splitDialer struct {
+	dialer     transport.StreamDialer
+	splitPoint int64
+}
+
+var _ transport.StreamDialer = (*splitDialer)(nil)
+
+// NewStreamDialer creates a client that splits the outgoing strean at byte splitpoint.
```
Comment (jyyi1) [inline transport/split/stream_dialer.go:31]: I think we need to clarify the exact byte index of `splitPoint`. Like if `splitPoint = 1`, does it mean `[0] |<-splitPoint [1] [2] ... ` or `[0] [1] |<-splitPoint [2] ... ` or something else?

```suggestion
// NewStreamDialer creates a [transport.StreamDialer] that splits the outgoing stream at byte splitPoint (starting from byte 1).
```

File: transport/split/writer.go [hunk 2]
```diff
@@ -0,0 +1,65 @@
...
+
+type SplitWriter struct {
+	writer     io.Writer
+	splitPoint int64
+}
+
+var _ io.Writer = (*SplitWriter)(nil)
+var _ io.ReaderFrom = (*SplitWriter)(nil)
+
+// NewWriter creates a Writer that ensures that the byte sequence is split at splitPoint,
```
Comment (jyyi1) [inline transport/split/writer.go:30]:
```suggestion
// NewWriter creates a [io.Writer] that ensures the byte sequence is split at splitPoint,
```

File: transport/split/stream_dialer.go (hunk 1 above)
Comment (fortuna) [reply transport/split/stream_dialer.go:31]: I renamed `splitPoint` to `prefixBytes` and updated the comment. I hope that helps.

File: README.md
Comment (jyyi1) [inline README.md:1]: is this file supposed to be included in this PR? It seems to be containing a lot of unrelated changes.

File: x/examples/http2transport/README.md
Comment (jyyi1) [inline x/examples/http2transport/README.md:1]: This entire folder seems to be another PR's change as well.

**Review by jyyi1 — APPROVED** [review]
Code LGTM, but please clean up the changes. They seem to be other PR's change.

**Comment by fortuna** [comment]
PR fixed!

File: transport/split/writer.go [hunk 3]
```diff
@@ -0,0 +1,66 @@
...
+		if err != nil {
+			return written, err
+		}
+	}
+	n, err := io.Copy(w.writer, source)
+	written += n
+	return written, err
+}
+
+func (w *SplitWriter) Write(data []byte) (written int, err error) {
```
Comment (ignoramous) [inline transport/split/writer.go:53]: The split causes issues with a few websites. For example, `jedynka.polskieradio.pl` refuses to complete the handshake ([ref](https://github.com/celzero/rethink-app/issues/964)).

We ended up resolving it using `outline-go-tun2socks`'s `Retrier` which first attempts the handshake as-is before attempting a `ClientHello` split [(ex)[https://github.com/Jigsaw-Code/outline-go-tun2socks/blob/dada2652ae2c6205f2daa3f88c805bbd6b28a713/intra/split/retrier.go#L120]).
